pkg/spanconfig/spanconfigsqlwatcher/sql_watcher.go

```diff
@@ -0,0 +1,453 @@
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/spanconfig"
+	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/lease"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/systemschema"
+	"github.com/cockroachdb/cockroach/pkg/sql/rowenc"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/log/logcrash"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/errors"
+)
+
+// SQLWatcher implements the spanconfig.SQLWatcher interface.
+var _ spanconfig.SQLWatcher = &SQLWatcher{}
+
+// The SQLWatcher is in charge of watching for specific SQL events and reacting
+// to them by constructing the list of span configurations implied as a result.
+// It establishes a rangefeed over system.zones and system.descriptor to do, as
+// changes to these tables may result in changes to the implied span
+// configurations for a tenant.
+//
+// When the SQLWatcher learns about a particular descriptor or zone
+// configuration  update, it constructs and sends the list of implied span
+// configurations on the channel returned by the Watch, which is the
+// orchestrating method for this whole process.
+type SQLWatcher struct {
+	codec            keys.SQLCodec
+	DB               *kv.DB
+	settings         *cluster.Settings
+	rangeFeedFactory *rangefeed.Factory
+	clock            *hlc.Clock
+	ie               sqlutil.InternalExecutor
+	leaseManager     *lease.Manager
+	stopper          *stop.Stopper
+	knobs            *spanconfig.TestingKnobs
+}
+
+// New constructs and returns a SQLWatcher.
+func New(
+	codec keys.SQLCodec,
+	DB *kv.DB,
+	settings *cluster.Settings,
+	rangeFeedFactory *rangefeed.Factory,
+	clock *hlc.Clock,
+	ie sqlutil.InternalExecutor,
+	leaseManager *lease.Manager,
+	stopper *stop.Stopper,
+	knobs *spanconfig.TestingKnobs,
+) *SQLWatcher {
+	if knobs == nil {
+		knobs = &spanconfig.TestingKnobs{}
+	}
+	return &SQLWatcher{
+		codec:            codec,
+		DB:               DB,
+		settings:         settings,
+		rangeFeedFactory: rangeFeedFactory,
+		clock:            clock,
+		ie:               ie,
+		leaseManager:     leaseManager,
+		stopper:          stopper,
+		knobs:            knobs,
+	}
+}
+
+// Watch is the entry point for the SQLWatcher.
+func (s *SQLWatcher) Watch(ctx context.Context) (<-chan spanconfig.Update, error) {
```

Comment:
Rename to WatchForSQLUpdates, the comment should be of the form "WatchForSQLUpdates is part of the spanconfig.SQLWatcher interface".